### ISSUE TITLE
Затыкание дыр в БЗ

### ISSUE THREAD

**Issue by TonitaN** [BaldiSlayer/rofl-lab1#79]
Когда вольётся основная масса данных из БЗ, придётся ещё заполнять смысловые промежутки между ними. Поскольку в этом варианте всё повеселее пока с LLM (в другом ещё нет подключения эмбеддинга вообще), с этим я помогу (залью несколько элементов в БЗ от себя), но и другие могут посильно включаться. С мерой близости ещё тоже нужно доразобраться (много лишнего в контексте, из-за чего иногда бывает ухудшение), и подумать про ключевые слова, но уже ориентиры на реальные проблемы в БЗ хотя бы минимальные появились.

**Comment by TonitaN** [comment]
Очень классно начат ответ, и видно, что контекст удачно сработал, но закончен, к сожалению, галлюцинацией:
https://gist.github.com/stewkk/60447a1872454c005ff3c5226a58996b
Т.е. тут очевидная дыра в БЗ.

Проблемы: нет ни одного подробного разбора аналогичной задачи. Нет подсказки про сохранение регулярной структуры.

**Comment by TonitaN** [comment]
Дыра в БЗ, особенно заметная в сочетании с нечёткой метрикой:
https://gist.github.com/stewkk/3a7fdeb81504e4d29e6c84d831b0b8b8
Нигде не сказали, что лемма о накачке не является достаточным условием КС-свойства. И ответ с пересечением тоже явно требует развёртки.

**Comment by TonitaN** [comment]
Нужна тонкая настройка, чтобы LLM меньше путал регулярные и нерегулярные языки в таких примерах. В частности, примеры с точно регулярными:
https://gist.github.com/stewkk/cca74585dc76ca912da5b6a18409d6d7

**Comment by TonitaN** [comment]
Тут какие-то жёсткие галлюцинации. 100500 раз надо повторить, что регулярные языки всегда КС, и как-то ещё донести, что такие примеры не подходят под регулярные (разбор a^n b^n явно?).
https://gist.github.com/stewkk/b1e7ce454c8fb733c7559f9bf16130ae

**Comment by TonitaN** [comment]
Зияющая дыра, которую пока никто даже не пробовал затыкать:
https://gist.github.com/stewkk/dec4c48ca845dc6e6e20948eae14f7eb

**Comment by TonitaN** [comment]
Очень коварная галлюцинация, которая в реальности будет сбивать с толку новичков:
https://gist.github.com/stewkk/bb024f603b30d4cf4b910dbe812240e8

**Comment by TonitaN** [comment]
Это гроб, и по жизни трудно с такими терминами разбираться, а уж в LLM...

https://gist.github.com/stewkk/b7cbc4b98bb317c7f28e99fdbda5aa67

**Comment by TonitaN** [comment]
А вот тут единогласно галлюцинируем:
https://gist.github.com/stewkk/12b2b89098ccece35bee87f03403294c

**Comment by TonitaN** [comment]
Тут одинаково плохо всё - нужно больше контекстов про иерархию языков
https://gist.github.com/stewkk/5c1cb414730a86d792f759045c55a5f1

**Comment by TonitaN** [comment]
Тут видно, какими запросами нужно пополнять БЗ ещё
https://gist.github.com/stewkk/a746ae7f44e8ee7e72907def088d0044
В нынешнем контексте LLM не различает понятия эквивалентности множеств и выразимости одного через другое.

**Comment by TonitaN** [comment]
Магический язык Дика
https://gist.github.com/stewkk/401ae0681a1c1f3ab728f72a0948acbd

**Comment by TonitaN** [comment]
Мало общепринятых сокращений, только один элемент БЗ упоминает PDA
https://gist.github.com/stewkk/e6c3cc0a0a64dacf49eaec57072c0510
Если будет присоединение ключевых слов, то и не нужно, но похоже, в таких случаях можно обойтись и без них, если расширить БЗ с упоминанием соответствующих терминов

**Comment by TonitaN** [comment]
@molivka , FYI

https://gist.github.com/stewkk/7e9717373b2caa30abea7c1b218ced2a

https://gist.github.com/stewkk/eba3019c0f3818716ab12b3d62d7a074

https://gist.github.com/stewkk/5478a75d41d800876104b46270a21fbf

**Comment by TonitaN** [comment]
Не знаю, кого тегать:
https://gist.github.com/stewkk/a7c44fd3d3b4183fe55e6c0cd2f142dc
https://gist.github.com/stewkk/527f03012812e1180605896c1a12d42c
https://gist.github.com/stewkk/2aabe0c74f99003fd5935a75ace62faf

И сложное (нетривиальная галлюцинация) на эту же тему
https://gist.github.com/stewkk/cf3614d662ee21765aaed15d58ed0527

**Comment by BaldiSlayer** [comment]
> Не знаю, кого тегать (@HumsterProgrammer , @molivka ? мб ещё кто-нибудь, кто пишет про LL) https://gist.github.com/stewkk/a7c44fd3d3b4183fe55e6c0cd2f142dc https://gist.github.com/stewkk/527f03012812e1180605896c1a12d42c https://gist.github.com/stewkk/2aabe0c74f99003fd5935a75ace62faf
> 
> И сложное (нетривиальная галлюцинация) на эту же тему https://gist.github.com/stewkk/cf3614d662ee21765aaed15d58ed0527

интересно, смотрит ли сюда вообще кто-то, кто заполняет БЗ....

**Comment by molivka** [comment]
> интересно, смотрит ли сюда вообще кто-то, кто заполняет БЗ....

посчитаем конверсию из просмотра репозитория в этот ишью без твоего тега в общий чат тг и без?)

**Comment by BaldiSlayer** [comment]
> > интересно, смотрит ли сюда вообще кто-то, кто заполняет БЗ....
> 
> посчитаем конверсию из просмотра репозитория в этот ишью без твоего тега в общий чат тг и с?)

мне уже начинает надоедать тегать ес честно

**Comment by BaldiSlayer** [comment]
@TonitaN а у Вас случайно уже не было примеров, когда модель с контекстом отвечает "я не знаю", но при этом без контекста норм отвечает?

если есть - можете поделиться пожалуйста?

P.S.

Так происходит из моего [ограничения](https://github.com/BaldiSlayer/rofl-lab1/blob/1b4dcda4d7353002827f6ac2f74883ac26812674/internal/app/mclient/ctemplate.go#L6) в доп промпте). Я хочу попробовать фиксануть это тем, что если моделька отвечает сама - то заставить ее написать, что "При ответе база знаний не была использована"

Я боюсь столкнуться с тем, что она никогда не будет использовать бз))
либо не будет использовать бз, но писать об этом не будет, так что хочу тестов чуть побольше, у меня сейчас пока 2-3 штуки)

Вообще есть второй вариант - добавлять в базу знаний даже тупые вопросы, но чет не знаю, мне такой подход меньше нравится, непонятно зач нам тогда умная, знающая модель

**Comment by BaldiSlayer** [comment]
> "языков L_d" 

там сейчас еще наверное LR и обращение языка L^R начнет путаться )))

> Не видно, чтобы промт являлся источником таких проблем (как минимум, с моей точки зрения

на самом деле [ловили](https://gist.github.com/stewkk/160fcb3919e179d7fa4bbe455aed168d), когда промпт чуть ухудшал. Тут нужно подумать над тем, хотим ли мы давать отвечать модели без контекста на что-либо, или вообще нет. Я вроде чуть пошаманил с промптом и добился того, что он когда отвечает сам - генерит бред, но пишет перед бредом, что БЗ не использовал. А когда надо - использует БЗ, но опять же, прям отвечать за детерминированность - не берусь.

> Что сбивает с толку - у нас большая проблема с близкими:

поиск близких да, хотелось бы пофиксить, если в итоге никто этого вдруг не сделает, то я возьмусь тогда за это сам..  сейчас просто к сожалению есть что сдавать (а еще я хочу поделать другие лабы по ТФЯ, мне же надо как-то допуск получить)
~хотяяяя, можно не получать допуск, тогда будет целый академ на фикс поиска близких, как раз исследование проведу~

**Comment by TonitaN** [comment]
> 
> на самом деле [ловили](https://gist.github.com/stewkk/160fcb3919e179d7fa4bbe455aed168d), когда промпт чуть ухудшал. 
Тут в п.1 без контекста написан, увы, бред (требование такое, что под него подходят только грамматики, не распознающие вообще ничего). Поэтому говорить о прямо улучшении трудно, тем более, что ключевые критерии (беспрефиксность и детерминированность) вовсе не упомянуты.

**Comment by TonitaN** [comment]
Тут есть дыры, но мне нравится, в какую сторону всё движется:

https://gist.github.com/stewkk/74f7e93be293e3f21f589561db5c0f7e (есть тонкая неточность, указывающая на дыру в БЗ: ловушка в DPDA определяется не только переходами по буквам)
https://gist.github.com/stewkk/2d7a335033f91ab9cb2b92da8bd147e9 (автоперевод КС-языка в контексте доставил )

**Comment by TonitaN** [comment]
> 
> поиск близких да, хотелось бы пофиксить, если в итоге никто этого вдруг не сделает, то я возьмусь тогда за это сам.. 

Не надо, пожалуйста. Есть подозрение, что это даже не долго, если делать по-рабоче-крестьянски, без NDA, и без желания запушить идеальный вариант фикса, а только с желанием улучшить до удовлетворительного. Но у нас что-то получается, что у народа либо NDA-рисерч, либо в перспективе исследование длиной в академ...

**Comment by molivka** [comment]
кому продать душу, чтобы можно было безболезненно скипать сон и делать эту лабу...

**Comment by TonitaN** [comment]
Ахаха, в продажу души мы умеем. Го на практику в Переславль? ~ладно, ладно, больше не буду предлагать, и так уже вся подслушка боится что-либо писать про иу9 из-за этой практики~

**Comment by BaldiSlayer** [comment]
> кому продать душу, чтобы можно было безболезненно скипать сон и делать эту лабу...

им.

![image](https://github.com/user-attachments/assets/db5545d6-7585-4546-84a5-b1c66c58c1f7)

**Comment by BaldiSlayer** [comment]
> либо в перспективе исследование длиной в академ...

я так сказал, потому что я если честно не знаю скок у меня это займет, никогда таким не занимался))

**Comment by BaldiSlayer** [comment]
https://gist.github.com/stewkk/cda997397b2fcc7dcdd10df0521dae0c

Галю

**Comment by HumsterProgrammer** [comment]
https://gist.github.com/stewkk/da345782f17d3dbc11e3d62d40d1c920

**Comment by BaldiSlayer** [comment]
> https://gist.github.com/stewkk/da345782f17d3dbc11e3d62d40d1c920

Мда, она отвечает такое чувство не с использованием контекста, а контекстом))
Прикол в том, что я её сильно на него ориентировал

**Comment by TonitaN** [comment]
Ну тут есть нюанс с ПСП (сокращение такое отсутствует в вопросах БЗ).
https://gist.github.com/stewkk/a2758648bc20729a7d6782bbc1f0c1df
Тут опять - сокращение DCFL не употребляется
https://gist.github.com/stewkk/6843bfe2a82e14b0799900d0b0b77263

А вот это уже проблемно
https://gist.github.com/stewkk/893492bee71c909ab03d1a23eb2013e7

**Comment by BaldiSlayer** [comment]
https://gist.github.com/stewkk/cd2e94b094913c84418f52debb28973e

кажется я его смог победить...

**Comment by TonitaN** [comment]
Неприятная галлюцинация в самом конце ответа, а так всё хорошо начиналось:
https://gist.github.com/stewkk/1c65bbb1432fdc5d53e84afb8658cf14

Думаю, нужно дополнять БЗ конкретно теоремой М-Н для VPL

**Comment by BaldiSlayer** [comment]
> Неприятная галлюцинация в самом конце ответа, а так всё хорошо начиналось: https://gist.github.com/stewkk/1c65bbb1432fdc5d53e84afb8658cf14
> 
> Думаю, нужно дополнять БЗ конкретно теоремой М-Н для VPL

и надо видимо вот тут https://github.com/BaldiSlayer/rofl-lab1/blob/21609cddf9686bcf9169fd71cf6ab63eef2013bf/data/data.yaml#L1596 уточинить, что доказательство то для регулярных языков)

**Comment by TonitaN** [comment]
А вот тут странно:
https://gist.github.com/stewkk/70ad460a9ae2396f24508e671aff6853
Вообще проигнорировано уточнение про КС в запросе.

**Comment by BaldiSlayer** [comment]
> А вот тут странно: https://gist.github.com/stewkk/70ad460a9ae2396f24508e671aff6853 Вообще проигнорировано уточнение про КС в запросе.
> 
> Кстати, в ответ id: 165 нужно ещё добавить итерацию.

кажется, что он не умеет в `КС ~ контекстно-свободные`, потому что вот 

https://gist.github.com/stewkk/a3a262832ad58774d89498e5ff642abe

**Comment by BaldiSlayer** [comment]
блин пойти что ли сделать массив `replace`'ов и привести КС к контекстно-свободные...

**Comment by TonitaN** [comment]
Массив синонимов (по корням) - хорошая мысль. Но можно их докидывать, а не заменять. Иначе DCFL и языки DPDA тоже оказываются не видимы друг их друга, как и ПСП с Диком.

**Comment by TonitaN** [comment]
А вот на эту тему можно воткнуть элемент БЗ, если БЗ ещё кто-нибудь занимается
https://gist.github.com/stewkk/705325e872cbff9fa26503d67df37cd5

**Comment by TonitaN** [comment]
Мало про недетерминированность и детерминированность КС языков (похожие галлюцинации были выше, подчёркиваю, что они никуда не делись):

https://gist.github.com/stewkk/e5f30a1900cca4d14830d4479f7d05aa
https://gist.github.com/stewkk/3ec3c93c06e5eb06b3be94efe4b6f47f

**Comment by BaldiSlayer** [comment]
> Массив синонимов (по корням) - хорошая мысль. Но можно их докидывать, а не заменять. Иначе DCFL и языки DPDA тоже оказываются не видимы друг их друга, как и ПСП с Диком.

у меня сейчас появилась идея, что в принципе я сейчас добавил же несколько вопросов к одному ответу, тогда мы можем наугментировать так нам немного вопросов еще для каждого из ответов, через замену `контекстно-свободные` -> КС


P.S.

я добавил

```yaml
  - 'Относительно каких операций замкнуты КС языки?

    '
```

в вопрос https://github.com/BaldiSlayer/rofl-lab1/blob/21609cddf9686bcf9169fd71cf6ab63eef2013bf/data/data.yaml#L2953-L2964

и он теперь в контексте:

https://gist.github.com/BaldiSlayer/3d4864aef59d2a9fed92fde8e1561d27

**Comment by BaldiSlayer** [comment]
> > Массив синонимов (по корням) - хорошая мысль. Но можно их докидывать, а не заменять. Иначе DCFL и языки DPDA тоже оказываются не видимы друг их друга, как и ПСП с Диком.
> 
> у меня сейчас появилась идея, что в принципе я сейчас добавил же несколько вопросов к одному ответу, тогда мы можем наугментировать так нам немного вопросов еще для каждого из ответов, через замену `контекстно-свободные` -> КС
> 
> P.S.
> 
> я добавил
> 
> ```yaml
>   - 'Относительно каких операций замкнуты КС языки?
> 
>     '
> ```
> 
> в вопрос
> 
> https://github.com/BaldiSlayer/rofl-lab1/blob/21609cddf9686bcf9169fd71cf6ab63eef2013bf/data/data.yaml#L2953-L2964
> 
> и он теперь в контексте:
> 
> https://gist.github.com/BaldiSlayer/3d4864aef59d2a9fed92fde8e1561d27

PPS:

добавляю пока что у себя локально

**Comment by TonitaN** [comment]
Жёсткая галлюцинация:
https://gist.github.com/stewkk/cacf73b50936bbc7e7d3f05c367da875

**Comment by BaldiSlayer** [comment]
> у меня сейчас появилась идея, что в принципе я сейчас добавил же несколько вопросов к одному ответу, тогда мы можем наугментировать так нам немного вопросов еще для каждого из ответов, через замену контекстно-свободные -> КС

и не только, есть же и другие замены

регулярный -> автоматный например

**Comment by TonitaN** [comment]
АААААААААА
https://gist.github.com/stewkk/104ec11767819d170407c9bf90ae3932

**Comment by BaldiSlayer** [comment]
https://gist.github.com/BaldiSlayer/90967ca2de7dc9f3c5c92837e34ccc44

пупупу

**Comment by BaldiSlayer** [comment]
как я это сделал

**Comment by TonitaN** [comment]
Это очень странно:
https://gist.github.com/stewkk/1d4aff6f1e2b71652cf08b700e634219
Всё же видимо dynamic_k иногда имеет смысл, особенно когда в контекст попадает много близкого (1001 лемма о накачке)

**Comment by BaldiSlayer** [comment]
> Это очень странно: https://gist.github.com/stewkk/1d4aff6f1e2b71652cf08b700e634219 Всё же видимо dynamic_k иногда имеет смысл, особенно когда в контекст попадает много близкого (1001 лемма о накачке)

надо будет подумать...

**Comment by BaldiSlayer** [comment]
> > никаких разночтений
> 
> https://gist.github.com/stewkk/98784b74c0952c82ff1a642e0ca244e4 https://gist.github.com/stewkk/fbd4920df01338922541c0f7b0eefe35
> 
> Откуда этот метод Мириам? Похоже на "стандартную" галлюцинацию.

да, я вот независимо запустил у себя, задал боту вопрос и тоже получил в ответе метод Мириам, пока не понял откуда это

**Comment by TonitaN** [comment]
Дыра:
https://gist.github.com/stewkk/4d5202655487084cdcc50ff513804606

**Comment by TonitaN** [comment]
Причём смена накачки на насос лечит предыдущую галлюцинацию:
https://gist.github.com/stewkk/50fed8299187da6b17a6c1257441694d

**Comment by BaldiSlayer** [comment]
> Причём смена накачки на насос лечит предыдущую галлюцинацию: https://gist.github.com/stewkk/50fed8299187da6b17a6c1257441694d

наверное просто лемма Огдена не взлезает в контекст (я чуть ограничил его размер), а лемма о накачке находится повыше, потому что типа и в элементах, и в вопросе она есть, а на слово "без" внимание не обращается 😕

**Comment by TonitaN** [comment]
Пока по ощущениям получается довольно не устойчиво к малым изменениям запроса. Если не костылить, но сосредоточиться именно на БЗ, а не тесте изменения размера контекста и близостей, то нужно будет объединить синонимичные элементы БЗ типа 1001 теоремы Клини.

**Comment by TonitaN** [comment]
((
https://gist.github.com/stewkk/fb5f92c1e49ee406f91ae555f9aef38c

**Comment by BaldiSlayer** [comment]
> (( https://gist.github.com/stewkk/fb5f92c1e49ee406f91ae555f9aef38c
> 
> И потом ещё кто-то будет жаловаться, что галлюцинации плохо ищутся?

Так никто ж это не читает наверное =(
Проще же взять какой-то гроб по типу контекстно-свободных графовых грамматик (я их не знаю, просто я подумал, что если есть регулярные древесные, то почему бы не быть кс графовых), которые точно никто не возьмет и написать по ним элементы базы знаний

**Comment by TonitaN** [comment]
> 
> Проще же взять какой-то гроб по типу контекстно-свободных графовых грамматик (я их не знаю, просто я подумал, что если есть регулярные древесные, то почему бы не быть кс графовых), которые точно никто не возьмет и написать по ним элементы базы знаний

Вот с одной стороны фейспалм, а с другой стороны - предчувствую, что эти гробы растянут экзамен по тфя так, что опять придётся заполнять рапорт о нарушении режима...

**Comment by BaldiSlayer** [comment]
> (( https://gist.github.com/stewkk/fb5f92c1e49ee406f91ae555f9aef38c
> 
> UPD: НО КАК? https://gist.github.com/stewkk/77510352b5f071765fe0b58c4e3e5faf

Магия

**Comment by TonitaN** [comment]
Метод Мириам:
https://gist.github.com/stewkk/8758b448c3e1258aa3d4daf823a52f21

КАК ОНО ПОПАЛО В МЕТОДЫ ПРОВЕРКИ РЕГУЛЯРНОСТИ ЯЗЫКА? Мы пропустили какие-то ещё нюансы автоперевода?

**Comment by BaldiSlayer** [comment]
> Метод Мириам: https://gist.github.com/stewkk/8758b448c3e1258aa3d4daf823a52f21
> 
> КАК ОНО ПОПАЛО В МЕТОДЫ ПРОВЕРКИ РЕГУЛЯРНОСТИ ЯЗЫКА? Мы пропустили какие-то ещё нюансы автоперевода?

Кстати у нас теперь нет перевода при построении эмбеддингов, у нас теперь другой sentence-transformer....

Надо подумать, мб тут уже дело в том, на каком языке модель работает (я ес честно 0 в ml, надо учить матчасть идти)

**Comment by BaldiSlayer** [comment]
> Метод Мириам: https://gist.github.com/stewkk/8758b448c3e1258aa3d4daf823a52f21
> 
> КАК ОНО ПОПАЛО В МЕТОДЫ ПРОВЕРКИ РЕГУЛЯРНОСТИ ЯЗЫКА? Мы пропустили какие-то ещё нюансы автоперевода?

А тут в промпте нет слова "языка", поэтому пошёл регулярность других событий проверять))

**Comment by TonitaN** [comment]
Ну вот как он вытащил именно этот неправильный пример? В контексте же нормально всё
https://gist.github.com/stewkk/414c38d0332970bc4f6892693c7fe668

**Comment by TonitaN** [comment]
Так, всё-таки наследие магистра Йоды осталось в целых двух экземплярах:
https://gist.github.com/stewkk/9cb48cb333da1b8bf2f6ff44f06bb230
А вот про преобразование регулярки в НКА нормального элемента БЗ нет.

**Comment by TonitaN** [comment]
Много всего про бисимуляцию, и нигде не сказано, что это обобщение метода минимизации на НКА:
https://gist.github.com/stewkk/b1ee7836fe694d5c6c662f435bce8303
(Да, запрос очень дурацкий, знаю)

**Comment by TonitaN** [comment]
@BaldiSlayer , вот это #165 - убийца чатбота. У нас же так мало про накачки сейчас)))

**Comment by BaldiSlayer** [comment]
> @BaldiSlayer , вот это #165 - убийца чатбота. У нас же так мало про накачки сейчас)))

Жаль, что нет плачущего смайлика для реакций. Он бы мне очень пригодился

**Comment by TonitaN** [comment]
Хм...
https://gist.github.com/stewkk/55b572f920f0709b744c85deb395aa7e

**Comment by TonitaN** [comment]
Есть AFA, но не видна их связь с НКА
https://gist.github.com/stewkk/680c09e90a0a28b40332b04d3b609524

**Comment by TonitaN** [comment]
Я никого не тегаю, кто надо, сам заберёт галлюцинации.

**Comment by TonitaN** [comment]
https://gist.github.com/stewkk/8ab97d5027e8439af9aa1cbb6c583994

(может стать ещё хуже после вливания в БЗ инфы про LR-позиционные-автоматы)
https://gist.github.com/stewkk/87ee09548844b54cb7d3ea333b73827c

неаккуратности в определениях конгруэнций по контексту и нет подвязки к бисимуляции
https://gist.github.com/stewkk/cb5298a23b67c0e5ea38fb6e46036430

а вот тут мало попало в контекст, из-за чего ответ неверный
https://gist.github.com/stewkk/2477e02542b0e2c69f973aa013315c90

и причина этого видна тут - почему то вообще не попало ничего в контекст накачки
https://gist.github.com/stewkk/79e4e5f3b0983f4b25d24e2ad73f869b

**Comment by TonitaN** [comment]
Интересно, кто-нибудь всё ещё занимается БЗ в этом варианте?

**Comment by BaldiSlayer** [comment]
> Интересно, кто-нибудь всё ещё занимается БЗ в этом варианте?

Вроде как есть человек, кто хочет заняться затыканием дыр)

**Comment by TonitaN** [comment]
Сегодня ночью хочу отревьюить все PR-ы на эту тему, чтобы в пятницу ещё было время у всех, кто ещё жив на 5 лр, потестировать влитое и дозаткнуть. Завтра уже поздно будет (и скорее всего завтра ночью я не буду ничего ревьюить)

**Comment by molivka** [comment]
> Интересно, кто-нибудь всё ещё занимается БЗ в этом варианте?

да, но из-за курсача пришлось поставить всё в жизни на паузу...

**Comment by BaldiSlayer** [comment]
> Сегодня ночью хочу отревьюить все PR-ы на эту тему, чтобы в пятницу ещё было время у всех, кто ещё жив на 5 лр, потестировать влитое и дозаткнуть. Завтра уже поздно будет (и скорее всего завтра ночью я не буду ничего ревьюить)
> 
> Так что если кто-то почти готов, к ночи жду.

ночью это с 12 же?

**Comment by TonitaN** [comment]
С 12 ещё вечер. С 1 ночи 100%

**Comment by TonitaN** [comment]
Это всё, или у кого-то ещё вечер?

**Comment by HermanDp45** [comment]
На самом деле выглядит все очень грустно, насколько я понимал, дедлайн был до конца декабря, но, видимо, что-то не так понял, сейчас активно делаю, но понимаю, что если и успею, то только к самому утру, 
соответственно хотелось бы понять, есть ли еще возможность выполнить 5ю лр, потому что даже и не представляю, как жить после понимания того, что столько баллов будет утеряно

**Comment by TonitaN** [comment]
Нет, до конца декабря по лабе индивидуально. Просто сейчас есть шанс усилить чатбот и получить в перспективе больше допбаллов командой.

**Comment by HermanDp45** [comment]
Отлично. Тогда испортить случайно то, что есть сейчас не хочется) 
Но радует, что возможность все же сдать л.р. присутствует, спасибо!

**Comment by TonitaN** [comment]
Тогда на сегодня проверки вливаний всё. Надеюсь, БЗ удастся усилить к фарме.

**Comment by TonitaN** [comment]
@BaldiSlayer , теперь справедливость в вашем отношении восстановлена и ревью есть.
